AI #1929: Add supported features and appendix content for commitment program eligibility details (#2025)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/supported_features/commitment_program_eligibility_details.md

Reviewer: @Matt-Cowsert 
Keep an eye on #1882... I've recommended a change to the name of `Type`, which feels too overloaded of a term.  If @SanjnaSrivatsaVMware accepts my suggestion of changing it to `ProgramType`, you'd need to globally apply that change here.

Author: Applied. `ProgramType` now used throughout: naming conventions heading (line 9), requirement bullets (lines 13-14), provider table header (line 19), SQL JSON paths (lines 67, 78, 119, 137), and SQL alias renamed to `EligibleProgramType` (lines 67, 119).

Reviewer: @Matt-Cowsert 
Lots of changes in #1882 mean that much of the language in this PR needs to change.  Have a full editorial pass through all the English here and make sure it aligns with #1882.

Author: Done. Commit 5d5cad2 is a full structural rewrite aligned with #1882's current state:

- JSON structure description rewritten (line 7): `CommitmentPrograms` array with `ProgramType`, single-array model
- All three SQL queries updated (lines 63-79, 90-108, 116-139): `$.CommitmentPrograms` and `$.ProgramType`
- Glossary linking applied throughout (lines 5, 7, 11, 14-15, 53, 57, 59, 84, 86, 112, 114)
- Dataset reference added (line 5)
- "line item" replaced with "row" (line 14)
- Obsolete CapacityReservationTypes narrative removed (was line 17, replaced by single-array explanation on line 7)
- SQL preamble note rewritten (line 53)
- Query 3 note: removed UNION sentence referencing CapacityReservationTypes (line 114)

Reviewer: Confirmed.  Resolving.

Reviewer: Please tidy up the linter issues.

Author: complete, resolving

Note: Unused commitment rows (CommitmentDiscountStatus = "Unused") have CommitmentDiscountId populated and are included in CoveredCost. Practitioners seeking a utilization-adjusted rate should additionally filter on CommitmentDiscountStatus = "Used". Whether CommitmentEligibilityDetails MUST be populated on Unused rows is not explicitly addressed in the column requirements. If a provider does not populate it on Unused rows, the denominator excludes them while the numerator does not, potentially inflating the coverage rate.

```sql
SELECT
  CU.ServiceProviderName,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NOT NULL
    THEN CU.EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(CU.EffectiveCost) AS TotalEligibleCost,
  SAFE_DIVIDE(
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END),
    SUM(CU.EffectiveCost)
  ) AS CommitmentCoverageRate
FROM focus_data_table CU
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
  AND CU.CommitmentEligibilityDetails IS NOT NULL
GROUP BY CU.ServiceProviderName

> ```

Reviewer: This example needs massaging in some way -- but I'm not yet sure of the intent of the example.

Is the point to say, "this could have been covered by _anything_"?  Commitment Discount, Capacity Reservation, Other?  Feels like we need to be more specific, to check if a Commitment Discount has been specified.  But that depends on the intent of the use case in this section, which I'm not totally sure of.  (Thus the suggestion below, which may or may not be hitting the mark.)  

I'll defer to TF-3 to discuss and resolve the proper next step here.

```suggestion
SELECT
  CU.ServiceProviderName,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NOT NULL
    THEN CU.EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NOT NULL 
      OR JSON_EXTRACT(CU.CommitmentEligibilityDetails, '$.CommitmentDiscountTypes') IS NOT NULL
    THEN CU.EffectiveCost ELSE 0 END) AS TotalEligibleCost,
  SAFE_DIVIDE(
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END),
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL 
        OR JSON_EXTRACT(CU.CommitmentEligibilityDetails, '$.CommitmentDiscountTypes') IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END)
  ) AS CommitmentCoverageRate
FROM focus_data_table CU
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
GROUP BY CU.ServiceProviderName
```

Reviewer: _If_ you were to go with this suggestion, here is an AI-gen stab at the kind of additions you'd need to make to the preamble:

> This query computes a commitment discount coverage rate using only eligible charges as the denominator. Without eligibility data, practitioners typically divide covered spend by total spend, which produces a coverage rate that includes ineligible charges (e.g., storage services, support fees) in the denominator and may not reflect the actionable coverage opportunity.
> 
> By extracting the CommitmentDiscountTypes key from the JSON object, the denominator includes only charges that could realistically be covered by a financial discount. It deliberately ignores programs that only reserve capacity.
> 
> Note: Unused commitment rows (CommitmentDiscountStatus = "Unused") have CommitmentDiscountId populated and are included in CoveredCost. Practitioners seeking a utilization-adjusted rate should additionally filter the numerator to CommitmentDiscountStatus = "Used". Furthermore, this query uses an OR condition in the denominator to ensure that all covered spend is counted as eligible, safeguarding against scenarios where a provider might omit the eligibility JSON on already-covered or unused rows.

Author: Thanks for the suggestion. The approach here diverges from the current query design, and the SQL references the old `CommitmentDiscountTypes` path which no longer exists after the #1882 restructuring.

The current query filters on `CommitmentEligibilityDetails IS NOT NULL` (line 106), which achieves the same intent: the denominator only includes eligible charges. The note on line 88 explicitly flags the unused-row edge case you raised in the follow-up comment.

Happy to discuss in TF-3 if you think the OR-based approach is still preferable. Deferring to group on the right query pattern here.

Author: Correction to my earlier reply: adopted your OR-based denominator approach in commit 5c531b3, adapted for the `CommitmentPrograms` structure.

The original `IS NOT NULL` filter had a gap: if a provider omits CommitmentEligibilityDetails on already-covered rows, those rows drop out of both the numerator and denominator, producing a 0% coverage rate. A charge cannot be covered by a commitment program it is not eligible for, so treating covered rows as eligible by definition is logically sound.

Changes:
- SQL (line 96): denominator uses `CommitmentDiscountId IS NOT NULL OR CommitmentEligibilityDetails IS NOT NULL`
- WHERE clause no longer filters on `CommitmentEligibilityDetails IS NOT NULL` (query runs on all Usage rows)
- Preamble (line 86): explains the OR logic and the provider-omission edge case
- Unused-row note (line 88): updated to reflect that unused rows are now included in both numerator and denominator via CommitmentDiscountId

Reviewer: Looks good.  Resolving.

Note: When a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Costs are not deduplicated across types, since each type represents an independent purchasing opportunity.

```sql
SELECT
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType') AS EligibleProgramType,
  SUM(CU.BilledCost) AS TotalEligibleUncoveredCost
FROM focus_data_table CU
CROSS JOIN
  UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibilityDetails, '$.CommitmentPrograms')) AS CP
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
  AND CU.CommitmentDiscountId IS NULL
  AND CU.CommitmentProgramEligibilityDetails IS NOT NULL
  -- Replace with provider-specific discount-bearing program types
  AND JSON_VALUE(CP, '$.ProgramType') IN ('FlexibleSpendPlan', 'ResourceReservation')
GROUP BY
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType')

> ORDER BY TotalEligibleUncoveredCost DESC

Reviewer: As rationalized [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882/files#r3037180715), please drop the word Total from the query.

```suggestion
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType') AS EligibleProgramType,
  SUM(CU.BilledCost) AS EligibleUncoveredCost
FROM focus_data_table CU
CROSS JOIN
  UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibilityDetails, '$.CommitmentPrograms')) AS CP
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
  AND CU.CommitmentDiscountId IS NULL
  AND CU.CommitmentProgramEligibilityDetails IS NOT NULL
  -- Replace with provider-specific discount-bearing program types
  AND JSON_VALUE(CP, '$.ProgramType') IN ('FlexibleSpendPlan', 'ResourceReservation')
GROUP BY
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType')
ORDER BY EligibleUncoveredCost DESC
```

```

### Calculate Commitment Discount Coverage Rate with Eligibility-Adjusted Denominator

> 
> This query computes a [*commitment discount*](#glossary:commitment-discount) coverage rate using only eligible [*charges*](#glossary:charge) as the denominator. Without eligibility data, practitioners typically divide covered spend by total spend, which produces a coverage rate that includes ineligible charges (e.g., storage services, support fees) in the denominator and may not reflect the actionable coverage opportunity. This query targets discount-bearing programs only; for [*capacity reservation*](#glossary:capacity-reservation) utilization, see the capacity reservation query below.
> 
> Note: Unused commitment rows (CommitmentDiscountStatus = "Unused") have CommitmentDiscountId populated and will artificially inflate both the numerator and the denominator if left in the dataset. To calculate a true, utilization-adjusted coverage rate, practitioners should additionally filter out these rows (e.g., AND CU.CommitmentDiscountStatus != 'Unused').
> 
> ```sql
> WITH CommitmentDiscountEligible AS (
>   SELECT
>     CU.ServiceProviderName,
>     CU.EffectiveCost,
>     CU.CommitmentDiscountId
>   FROM focus_data_table CU
>   CROSS JOIN
>     UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibilityDetails, '$.CommitmentPrograms')) AS CP
>   WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
>     AND CU.ChargeCategory = 'Usage'
>     AND CU.CommitmentProgramEligibilityDetails IS NOT NULL
>     -- Replace with provider-specific discount-bearing program types
>     AND JSON_VALUE(CP, '$.ProgramType') IN ('FlexibleSpendPlan', 'ResourceReservation')
> )
> SELECT
>   ServiceProviderName,
>   SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END) AS CoveredCost,
>   SUM(EffectiveCost) AS TotalEligibleCost,
>   SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END)
>     / NULLIF(SUM(EffectiveCost), 0) AS CommitmentCoverageRate
> FROM CommitmentDiscountEligible
> GROUP BY ServiceProviderName
> ```

Reviewer: The query uses a `CROSS JOIN UNNEST` to unpack the `CommitmentPrograms` array. This is completely fine for Query 1 and Query 3 because they include `ProgramType` in their GROUP BY clauses.

However, Query 2 only groups by `ServiceProviderName`. If a usage row is eligible for both a FlexibleSpendPlan and a ResourceReservation (as shown in Row 1 and Row 2 of your `coverage_rate_eligibility_adjusted.csv` example), the `CROSS JOIN` will output two rows for that single charge. Because Query 2 doesn't group by program type or deduplicate, it will double-count the `EffectiveCost` for that charge, artificially inflating both your numerator (`CoveredCost`) and your denominator (`TotalEligibleCost`).

Additionally, we should implement the "Defensive OR" strategy mentioned in the column spec so we don't accidentally drop actively covered rows if a provider omits the eligibility JSON.

Here is the corrected SQL using an `EXISTS` subquery instead of a `CROSS JOIN`. This ensures each row is counted exactly once and properly handles missing JSON on covered rows:

```suggestion
```sql
WITH CommitmentDiscountEligible AS (
  SELECT
    CU.ServiceProviderName,
    CU.EffectiveCost,
    CU.CommitmentDiscountId
  FROM focus_data_table CU
  WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
    AND CU.ChargeCategory = 'Usage'
    AND (
      -- Defensive check: Always include actively covered rows in the denominator
      CU.CommitmentDiscountId IS NOT NULL
      -- If uncovered, check if the JSON array contains an eligible program type
      OR EXISTS (
        SELECT 1
        FROM UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibilityDetails, '$.CommitmentPrograms')) AS CP
        WHERE JSON_VALUE(CP, '$.ProgramType') IN ('FlexibleSpendPlan', 'ResourceReservation')
      )
    )
)
SELECT
  ServiceProviderName,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(EffectiveCost) AS TotalEligibleCost,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END)
    / NULLIF(SUM(EffectiveCost), 0) AS CommitmentCoverageRate
FROM CommitmentDiscountEligible
GROUP BY ServiceProviderName
```
```

### Compare Commitment Opportunities Across Providers (Cross-Provider with SaaS)

This query aggregates eligible spend and uncovered eligible spend across all providers, including SaaS platforms. A practitioner managing both CSP and SaaS workloads can identify where uncovered eligible spend is concentrated across [*commitment program*](#glossary:commitment-program) types.

> Note: As with the first query above, when a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Removing the EligibleProgramType grouping without deduplicating would inflate totals.

Reviewer: Take or leave this suggestion.  We already have a blanket disclaimer at the top about conformant artifacts; this would be doubling down on that disclaimer with a more specific note here.  Up to you whether you want to include it.

This query has an unavoidable limitation when dealing with non-compliant data generator data. Because this query unrolls the JSON array to group by `ProgramType`, if a data generator omits the JSON on a *covered* row, that row is entirely dropped by the `CROSS JOIN UNNEST`. We can't use the "Defensive OR" trick here because without the JSON, we don't know which `ProgramType` bucket to assign the covered spend to. 

Consequently, `TotalEligibleCost` will be underreported, which mathematically inflates the `UncoveredRate`. The SQL itself is completely sound for compliant data, but we should probably add a brief warning note above the query so practitioners understand why their rates might look skewed.

Here is a suggestion for a note to add right before the SQL block:

```suggestion
Notes: 

* As with the first query above, when a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Removing the EligibleProgramType grouping without deduplicating would inflate totals.
* If a data generator incorrectly omits `CommitmentProgramEligibilityDetails` on covered rows, those rows will be excluded from this query's output. Because the specific `ProgramType` cannot be determined without the JSON array, this missing data will artificially inflate the calculated `UncoveredRate` for the affected programs.

Author: I appreciate this point @shawnalpay and I initially had more "if the data generator doesn't do what we asking" type language in this section, but I dropped it and re-wrote the section to remove those, so I don't want to go back and add another one (though I acknowledge the point).

Reviewer: Understood.  Resolving.

> ```sql

Reviewer: In the introduction to Query 1, we explicitly state that we use `BilledCost` rather than `EffectiveCost` for calculating uncovered spend. 

> This query uses BilledCost rather than EffectiveCost because the charges are uncovered (CommitmentDiscountId IS NULL), so BilledCost reflects the actual amount paid...

However, in the CTE here, we switch to using `EffectiveCost` for that exact calculation (`CASE WHEN CU.CommitmentDiscountId IS NULL THEN CU.EffectiveCost ELSE 0 END AS UncoveredCost`).

Since BilledCost and EffectiveCost are identical on uncovered rows, the math works fine. But for the sake of the reader's mental model, should we swap `CU.EffectiveCost` to `CU.BilledCost` in that CASE statement so it matches the logic we explained in Query 1?

Author: Good catch. Switched UncoveredCost to use BilledCost for consistency with Query 1's rationale. For uncovered rows (CommitmentDiscountId IS NULL), BilledCost and EffectiveCost are equivalent, but BilledCost is the semantically correct choice per the reasoning in Query 1. EffectiveCost retained for the total and denominator since covered rows reflect discount pricing.

Reviewer: Confirmed.  Resolving.

CommitmentProgramEligibilityDetails contains a JSON object with a FOCUS-defined top-level key `CommitmentPrograms` containing an array of objects. Each object has a `ProgramType` property identifying the specific *commitment program*. Both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and capacity-reservation programs (e.g., Advance Resource Commitments) are included in the same array, distinguished by their `ProgramType` value. [*Service providers*](#glossary:service-provider) may include additional custom keys (prefixed with x_) to pass through extra metadata or provider-specific attributes related to the eligibility. Per the [column requirements](#datasets.costandusage.commitmentprogrameligibilitydetails), service providers may include negotiated *commitment programs* when the usage is eligible and the program is not broadly applicable across the service provider's service catalog. For more information, see [CommitmentProgramEligibilityDetails](#datasets.costandusage.commitmentprogrameligibilitydetails).

### Naming Conventions for ProgramType Values

> The `ProgramType` property follows PascalCase by convention, identifying [*commitment programs*](#glossary:commitment-program) supported by the provider. Per the [column requirements](#datasets.costandusage.commitmentprogrameligibilitydetails), these values:

Author: Good catch on PascalCase - linked that glossary term. However, the suggestion also strips italic from "commitment programs," which should remain `*commitment programs*` per editorial convention (subsequent glossary occurrences keep italic, just drop the link). Applied both fixes in commit 1d79b07e.

### Naming Conventions for ProgramType Values

The `ProgramType` property follows [*PascalCase*](#glossary:pascalcase) by convention, identifying *commitment programs* supported by the provider. Per the [column requirements](#datasets.costandusage.commitmentprogrameligibilitydetails), these values:

* Equals [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in the CommitmentPrograms array when CommitmentDiscountType is not null.
* Corresponds to terminology disclosed by the service provider in public documentation. This guidance is especially relevant for SaaS providers that do not itemize commitment discount application at the row level, where CommitmentDiscountType is typically not populated.
* Does not encode period length, payment option, or other commitment attributes (e.g., use "FlexibleSpendPlan" rather than "1YearFlexibleSpendPlanNoUpfront").

> * Uses the provider name as-is where a provider's documented program name inherently includes a period reference (e.g., StackLens's "MonthlyPlatformCommitment").

Reviewer: Need to resolve singular vs plural, given that the lead-in phrase is "these values":

```suggestion
* Equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in the CommitmentPrograms array when CommitmentDiscountType is not null.
* Correspond to terminology disclosed by the service provider in public documentation. This guidance is especially relevant for SaaS providers that do not itemize commitment discount application at the row level, where CommitmentDiscountType is typically not populated.
* Do not encode period length, payment option, or other commitment attributes (e.g., use "FlexibleSpendPlan" rather than "1YearFlexibleSpendPlanNoUpfront").
* Use the provider name as-is where a provider's documented program name inherently includes a period reference (e.g., StackLens's "MonthlyPlatformCommitment").
```

      )
    )
)
SELECT
  ServiceProviderName,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(EffectiveCost) AS EligibleCost,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END)
    / NULLIF(SUM(EffectiveCost), 0) AS CommitmentCoverageRate
FROM CommitmentDiscountEligible
GROUP BY ServiceProviderName
```

### Compare Commitment Opportunities Across Providers (Cross-Provider with SaaS)

This query aggregates eligible spend and uncovered eligible spend across all providers, including SaaS platforms. A practitioner managing both CSP and SaaS workloads can identify where uncovered eligible spend is concentrated across [*commitment program*](#glossary:commitment-program) types.

Note: As with the first query above, when a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Removing the EligibleProgramType grouping without deduplicating would inflate totals.

```sql
WITH CommitmentDiscountEligible AS (
  SELECT
    CU.ServiceProviderName,
    JSON_VALUE(CP, '$.ProgramType') AS ProgramType,
    CU.EffectiveCost,

>     CASE WHEN CU.CommitmentDiscountId IS NULL THEN CU.BilledCost ELSE 0 END AS UncoveredCost

Reviewer: In the CTE, `UncoveredCost` is defined using `BilledCost`:
```sql
CASE WHEN CU.CommitmentDiscountId IS NULL THEN CU.BilledCost ELSE 0 END AS UncoveredCost
```

Then, in the final `SELECT`, the ratio divides that `BilledCost`-derived numerator by an `EffectiveCost`-derived denominator:
```sql
SUM(UncoveredCost) / NULLIF(SUM(EffectiveCost), 0) AS UncoveredRate
```

Because you are filtering for uncovered rows (`CommitmentDiscountId IS NULL`), `BilledCost` and `EffectiveCost` are mathematically identical here. However, mixing the two column concepts in a single ratio might cause a bit of head-scratching for practitioners reading the spec. 

**Suggestion:** Change `BilledCost` to `EffectiveCost` in the CTE so the entire calculation is explicitly grounded in the same column logic:
```suggestion
    CASE WHEN CU.CommitmentDiscountId IS NULL THEN CU.EffectiveCost ELSE 0 END AS UncoveredCost
```

(Or align on just using `BilledCost` throughout.)

  SUM(UncoveredCost) / NULLIF(SUM(EffectiveCost), 0) AS UncoveredRate
FROM CommitmentDiscountEligible
GROUP BY ServiceProviderName, ProgramType
ORDER BY UncoveredEligibleCost DESC
```

### Identify Eligible Capacity Reservation Spend

[*Capacity reservations*](#glossary:capacity-reservation) secure resource availability rather than provide discounts, and are tracked via [CapacityReservationId](#datasets.costandusage.capacityreservationid) and [CapacityReservationStatus](#datasets.costandusage.capacityreservationstatus) rather than the [*commitment discount*](#glossary:commitment-discount) columns used in the queries above. This query identifies [*charges*](#glossary:charge) eligible for capacity-reservation [*commitment programs*](#glossary:commitment-program), distinguishing between used and unused reservations.

The query filters [CommitmentProgramEligibilityDetails](#datasets.costandusage.commitmentprogrameligibilitydetails) to rows whose `ProgramType` values correspond to capacity-reservation programs (e.g., "AdvanceResourceCommitment", "ZonalResourceCommitment"). It then uses CapacityReservationId and CapacityReservationStatus to determine reservation utilization.

> Note: The FOCUS specification requires CapacityReservationId to not be null when a charge represents unused capacity (MUST), but only recommends (SHOULD) populating it when a charge is related to a [*capacity reservation*](#glossary:capacity-reservation). Where a provider does not populate CapacityReservationId on used rows, this query will show those rows with a null CapacityReservationStatus.

Reviewer: Nit.

```suggestion
Note: The FOCUS specification requires CapacityReservationId to not be null when a charge represents unused capacity (MUST), but only recommends (SHOULD) populating it when a charge is related to a used [*capacity reservation*](#glossary:capacity-reservation). Where a provider does not populate CapacityReservationId on used rows, this query will show those rows with a null CapacityReservationStatus.
```

Author: Addressed in commit 41f1aa7. Removed the parentheticals entirely rather than lowercasing -- "requires" and "recommends" already carry the intended meaning without the BCP-14 keywords. Also fixed the glossary link-once violation on *capacity reservation* and changed "a provider" to "a data generator" for consistency with the spec's terminology.

```sql
SELECT
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType') AS EligibleProgramType,
  CU.CapacityReservationStatus,
  SUM(CU.BilledCost) AS BilledCost,
  COUNT(*) AS RowCount
FROM focus_data_table CU
CROSS JOIN
  UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibilityDetails, '$.CommitmentPrograms')) AS CP
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
  AND CU.CommitmentProgramEligibilityDetails IS NOT NULL
  AND JSON_VALUE(CP, '$.ProgramType') IN ('AdvanceResourceCommitment', 'ZonalResourceCommitment')
GROUP BY
  CU.ServiceProviderName,
  CU.ServiceName,
  JSON_VALUE(CP, '$.ProgramType'),
  CU.CapacityReservationStatus
ORDER BY BilledCost DESC
```

> ## Introduced (Version)

Reviewer: ```suggestion
## Introduced (version)
```

# Commitment Program Eligibility Details

## Description

> FOCUS supports the identification of [*charges*](#glossary:charge) in the [Cost and Usage](#datasets.costandusage) dataset that are eligible for [*commitment programs*](#glossary:commitment-program). The [CommitmentProgramEligibilityDetails](#datasets.costandusage.commitmentprogrameligibilitydetails) column captures which *commitment programs* a charge qualifies for, regardless of whether a [*commitment*](#glossary:commitment) is currently applied. This enables practitioners to calculate eligibility-adjusted commitment coverage rates, identify uncovered savings opportunities, and compare commitment options across providers.

Reviewer: According to the `AGENTS.md` guidelines, we need to use the display name rather than the PascalCase entity ID when writing non-normative, introductory text.

Per the guidelines:
> **Column/Attribute Display Names:** Normal text with spaces (e.g., "Pricing Quantity"). These SHOULD be used in introductory or explanatory non-normative text.

Currently, the text uses the PascalCase ID:
> `The [CommitmentProgramEligibilityDetails](#datasets.costandusage.commitmentprogrameligibilitydetails) column captures which...`

We should change this to use the display name "Commitment Program Eligibility Details" while keeping the hyperlink intact.

**Suggested fix:**
```suggestion
FOCUS supports the identification of [*charges*](#glossary:charge) in the [Cost and Usage](#datasets.costandusage) dataset that are eligible for [*commitment programs*](#glossary:commitment-program). The [Commitment Program Eligibility Details](#datasets.costandusage.commitmentprogrameligibilitydetails) column captures which *commitment programs* a charge qualifies for, regardless of whether a [*commitment*](#glossary:commitment) is currently applied. This enables practitioners to calculate eligibility-adjusted commitment coverage rates, identify uncovered savings opportunities, and compare commitment options across providers.
```

> CommitmentProgramEligibilityDetails contains a JSON object with a FOCUS-defined top-level key `CommitmentPrograms` containing an array of objects. Each object has a `ProgramType` property identifying the specific *commitment program*. Both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and capacity-reservation programs (e.g., Advance Resource Commitments) are included in the same array, distinguished by their `ProgramType` value. [*Data generators*](#glossary:data-generator) may include additional custom keys (prefixed with x_) to pass through extra metadata or provider-specific attributes related to the eligibility. Per the [column requirements](#datasets.costandusage.commitmentprogrameligibilitydetails), service providers may include negotiated *commitment programs* when the usage is eligible and the program is not broadly applicable across the service provider's service catalog. For more information, see the CommitmentProgramEligibilityDetails column definition.

Reviewer: We don't yet have a glossary entry for `data generator`.  (We likely will soon, but not yet.)

```suggestion
CommitmentProgramEligibilityDetails contains a JSON object with a FOCUS-defined top-level key `CommitmentPrograms` containing an array of objects. Each object has a `ProgramType` property identifying the specific *commitment program*. Both discount-bearing programs (e.g., Flexible Spend Plans, Resource Reservations) and capacity-reservation programs (e.g., Advance Resource Commitments) are included in the same array, distinguished by their `ProgramType` value. [*Data generators*](#metadata:datagenerator) may include additional custom keys (prefixed with x_) to pass through extra metadata or provider-specific attributes related to the eligibility. Per the [column requirements](#datasets.costandusage.commitmentprogrameligibilitydetails), service providers may include negotiated *commitment programs* when the usage is eligible and the program is not broadly applicable across the service provider's service catalog. For more information, see the CommitmentProgramEligibilityDetails column definition.
```

  SUM(UncoveredCost) / NULLIF(SUM(EffectiveCost), 0) AS UncoveredRate
FROM CommitmentDiscountEligible
GROUP BY ServiceProviderName, ProgramType
ORDER BY UncoveredEligibleCost DESC
```

### Identify Eligible Capacity Reservation Spend

[*Capacity reservations*](#glossary:capacity-reservation) secure resource availability rather than provide discounts, and are tracked via [CapacityReservationId](#datasets.costandusage.capacityreservationid) and [CapacityReservationStatus](#datasets.costandusage.capacityreservationstatus) rather than the [*commitment discount*](#glossary:commitment-discount) columns used in the queries above. This query identifies [*charges*](#glossary:charge) eligible for capacity-reservation [*commitment programs*](#glossary:commitment-program), distinguishing between used and unused reservations.

The query filters [CommitmentProgramEligibilityDetails](#datasets.costandusage.commitmentprogrameligibilitydetails) to rows whose `ProgramType` values correspond to capacity-reservation programs (e.g., "AdvanceResourceCommitment", "ZonalResourceCommitment"). It then uses CapacityReservationId and CapacityReservationStatus to determine reservation utilization.

> Note: The FOCUS specification requires CapacityReservationId to not be null when a charge represents unused capacity (MUST), but only recommends (SHOULD) populating it when a charge is related to a used [*capacity reservation*](#glossary:capacity-reservation). Where a provider does not populate CapacityReservationId on used rows, this query will show those rows with a null CapacityReservationStatus.

Reviewer: I'm gonna leave it to you to decide whether to accept this suggestion.  Just know that, if you don't, every LLM going forward will stumble over this because it fails the styleguide.

Also: need to strike the glossary reference -- not the first reference in the section.

---

According to the `AGENTS.md` styleguide, normative BCP-14 keywords (MUST, SHOULD, etc.) cannot leak into non-normative sections:

> "Normative keywords MUST ONLY appear under a 'Requirements' header. Flag any normative keywords leaking into 'Description', 'Examples', or other non-normative sections."

**Suggestion:** Lowercase the keywords or remove the parentheses entirely, since the words "requires" and "recommends" already convey the intended meaning. For example:

```suggestion
Note: The FOCUS specification requires CapacityReservationId to not be null when a charge represents unused capacity (must), but only recommends (should) populating it when a charge is related to a used *capacity reservation*. Where a provider does not populate CapacityReservationId on used rows, this query will show those rows with a null CapacityReservationStatus.
```

Author: Good catch on the styleguide violation. Addressed in commit 41f1aa7. Went with removing the parentheticals entirely rather than lowercasing -- "requires" and "recommends" already convey the normative intent without BCP-14 keywords. Also struck the glossary link on *capacity reservation* (not first reference) and changed "a provider" to "a data generator" per spec terminology.

```sql
SELECT
  CU.ServiceProviderName,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NOT NULL
    THEN CU.EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NOT NULL
      OR CU.CommitmentProgramEligibility IS NOT NULL
    THEN CU.EffectiveCost ELSE 0 END) AS TotalEligibleCost,
  SAFE_DIVIDE(
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END),
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL
        OR CU.CommitmentProgramEligibility IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END)
  ) AS CommitmentCoverageRate
FROM focus_data_table CU
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
GROUP BY CU.ServiceProviderName

> ```

Reviewer 2: **Consider this alternative:**

```sql
WITH DiscountEligible AS (
  SELECT
    CU.ServiceProviderName,
    CU.EffectiveCost,
    CU.CommitmentDiscountId
  FROM focus_data_table CU
  WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
    AND CU.ChargeCategory = 'Usage'
    AND EXISTS (
      SELECT 1
      FROM UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibility, '$.CommitmentPrograms')) AS CP
      WHERE JSON_VALUE(CP, '$.ProgramType') NOT IN ('AdvanceResourceCommitment', 'ZonalResourceCommitment')
    )
)
SELECT
  ServiceProviderName,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END) AS CoveredCost,
  SUM(EffectiveCost) AS TotalEligibleCost,
  SUM(CASE WHEN CommitmentDiscountId IS NOT NULL THEN EffectiveCost ELSE 0 END) / NULLIF(SUM(EffectiveCost), 0) AS CommitmentCoverageRate
FROM DiscountEligible
GROUP BY ServiceProviderName
```

**How the query works:**

1. CTE filters to Usage charges that are eligible for at least one discount-bearing program type
2. Main query groups by provider and calculates:
   - CoveredCost: sum of EffectiveCost where CommitmentDiscountId is populated
   - TotalEligibleCost: sum of all EffectiveCost (already filtered to discount-eligible)
   - CommitmentCoverageRate: ratio of covered to eligible (NULLIF prevents division by zero)

**This query:**
- Excludes known capacity-reservation program types (via NOT IN). Could alternatively use IN with known discount-bearing types (e.g., `IN ('FlexibleSpendPlan', 'ResourceReservation')`). Ideally, the specification would provide a way to distinguish discount-bearing vs capacity-reservation program types without hardcoding values.
- Assumes FOCUS-conformant data -  no workarounds needed. See [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2025/files#r3028432233) regarding adding a general conformance note at the top of the SQL queries section.
- Uses CTE + NULLIF for better ANSI SQL portability (both are SQL standard, unlike SAFE_DIVIDE).

**Note:** If adopting this approach, the query description text should be updated to reflect the filtering logic and removal of the OR workaround.

Author: Applied in commit 0ab0e94. Adopted the CTE structure and NULLIF. Used IN with known discount-bearing program types rather than NOT IN with capacity types. Reasoning: if a new program type is introduced and the query is not updated, IN produces a conservative undercount (safer failure mode) while NOT IN would inflate the coverage rate. Added a SQL comment guiding practitioners to substitute their provider-specific values.

Note: As with the first query above, when a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Removing the EligibleProgramType grouping without deduplicating would inflate totals.

```sql
SELECT
  CU.ServiceProviderName,
  JSON_VALUE(CP, '$.ProgramType') AS EligibleProgramType,
  SUM(CU.EffectiveCost) AS TotalEligibleCost,
  SUM(CASE
    WHEN CU.CommitmentDiscountId IS NULL
    THEN CU.EffectiveCost ELSE 0 END) AS UncoveredEligibleCost,
  SAFE_DIVIDE(
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NULL
      THEN CU.EffectiveCost ELSE 0 END),
    SUM(CU.EffectiveCost)
  ) AS UncoveredRate
FROM focus_data_table CU
CROSS JOIN
  UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibility, '$.CommitmentPrograms')) AS CP
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
  AND CU.CommitmentProgramEligibility IS NOT NULL
GROUP BY
  CU.ServiceProviderName,
  JSON_VALUE(CP, '$.ProgramType')
ORDER BY UncoveredEligibleCost DESC

> ```

Reviewer 2: **Consider this alternative:**

```sql
WITH DiscountEligible AS (
  SELECT
    CU.ServiceProviderName,
    JSON_VALUE(CP, '$.ProgramType') AS ProgramType,
    CU.EffectiveCost,
    CASE WHEN CU.CommitmentDiscountId IS NULL THEN CU.EffectiveCost ELSE 0 END AS UncoveredCost
  FROM focus_data_table CU
  CROSS JOIN
    UNNEST(JSON_EXTRACT_ARRAY(CU.CommitmentProgramEligibility, '$.CommitmentPrograms')) AS CP
  WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
    AND CU.ChargeCategory = 'Usage'
    AND CU.CommitmentProgramEligibility IS NOT NULL
    AND JSON_VALUE(CP, '$.ProgramType') NOT IN ('AdvanceResourceCommitment', 'ZonalResourceCommitment')
)
SELECT
  ServiceProviderName,
  ProgramType AS EligibleProgramType,
  SUM(EffectiveCost) AS TotalEligibleCost,
  SUM(UncoveredCost) AS UncoveredEligibleCost,
  SUM(UncoveredCost) / NULLIF(SUM(EffectiveCost), 0) AS UncoveredRate
FROM DiscountEligible
GROUP BY ServiceProviderName, ProgramType
ORDER BY UncoveredEligibleCost DESC
```

**How the query works:**
1. CTE expands CommitmentPrograms array and filters to discount-bearing program types only
2. Pre-calculates UncoveredCost (EffectiveCost where CommitmentDiscountId is null, else 0)
3. Main query groups by provider and program type, calculates:
   - TotalEligibleCost: sum of EffectiveCost per program type
   - UncoveredEligibleCost: sum of uncovered costs per program type
   - UncoveredRate: ratio of uncovered to eligible (NULLIF prevents division by zero)

**This query:**
- Excludes known capacity-reservation program types (via NOT IN). Could alternatively use IN with known discount-bearing types (e.g., `IN ('FlexibleSpendPlan', 'ResourceReservation')`). Ideally, the specification would provide a way to distinguish discount-bearing vs capacity-reservation program types without hardcoding values.
- Assumes FOCUS-conformant data — no workarounds needed. See Comment 4 regarding adding a general conformance note at the top of the SQL queries section.
- Uses CTE + NULLIF for better ANSI SQL portability (both are SQL standard, unlike SAFE_DIVIDE).
- CTE avoids repeating CASE logic and maintains consistency with Query 2 approach.

**Note:** If adopting this approach, the query description text should be updated to remove the SaaS provider non-conformance note.

**Minimal alternative:** If preferring to keep the original query structure, just add the NOT IN filter:
```sql
AND JSON_VALUE(CP, '$.ProgramType') NOT IN ('AdvanceResourceCommitment', 'ZonalResourceCommitment')
```

Author: Applied in commit 0ab0e94. Same approach as Query 2: CTE with UncoveredCost pre-calculated, NULLIF, IN filter with practitioner guidance comment.

      THEN CU.EffectiveCost ELSE 0 END),
    SUM(CASE
      WHEN CU.CommitmentDiscountId IS NOT NULL
        OR CU.CommitmentProgramEligibility IS NOT NULL
      THEN CU.EffectiveCost ELSE 0 END)
  ) AS CommitmentCoverageRate
FROM focus_data_table CU
WHERE CU.ChargePeriodStart >= ? AND CU.ChargePeriodEnd < ?
  AND CU.ChargeCategory = 'Usage'
GROUP BY CU.ServiceProviderName
```

### Compare Commitment Opportunities Across Providers (Cross-Provider with SaaS)

This query aggregates eligible spend and uncovered eligible spend across all providers, including SaaS platforms. A practitioner managing both CSP and SaaS workloads can identify where uncovered eligible spend is concentrated across [*commitment program*](#glossary:commitment-program) types.

Note: Some SaaS providers may not populate CommitmentDiscountId even when a [*commitment*](#glossary:commitment) is applied. For those providers, EffectiveCost may not reflect *commitment* pricing, and this query captures total eligible spend rather than distinguishing covered from uncovered. Practitioners should consult provider-specific documentation to determine actual *commitment* utilization.

> Note: As with the first query above, when a charge is eligible for multiple *commitment program* types, it appears once per eligible type. Removing the EligibleProgramType grouping without deduplicating would inflate totals.

Reviewer 3: No suggestion here, but it looks like we have a common theme emerging here of warning the user of possible inflation due to provider choices

Author: Agreed there was a theme. The conformance note added at the top of the SQL section (commit 0ab0e94) consolidates the pattern: queries assume conformant data, and the remaining inflation warnings are limited to the unused-row note on Query 2 (which documents a real query behavior, not a provider conformance issue).
